src/stack-unix/udpv4v6_socket.ml

```diff
@@ -191,6 +190,11 @@ let listen t ~port callback =
                            | Some v4 -> Ipaddr.V4 v4
                          in
                          let dst = Ipaddr.(V6 V6.unspecified) in (* TODO *)
+                         let buf =
+                           let b = Cstruct.create_unsafe len in
+                           Cstruct.blit buf 0 b 0 len;
+                           b
+                         in
                          callback ~src ~dst ~src_port buf
```

Comment:
What I understood from the report #501 is that here this callback in the Unix stack is run synchronously with Lwt_cstruct.recvfrom (so there won't be multiple callbacks executed concurrently) -- while in the Mirage stack due to every mirage-net-* implementation calling "Lwt.async" for the callback of each incoming packet, the MirageOS UDP callback is asynchronous (i.e. there can be multiple callbacks concurrently).

I still don't quite understand what the desired semantics should be -- maybe the callback should not involve "io / Lwt.t" at all? Maybe it should be all sync? My intuition is that the "Lwt.async" for each incoming packet is not optimal from the performance and scheduling perspective.

Reply:
I think this PR is orthogonal to #501. @rand00 also reported a buffer ownership bug on IRC (but no issue AFAICT).

Reply:
ah, ok :)